opensearch_single_kernel/managers/lock.py

```diff
@@ -0,0 +1,507 @@
+2. if there are no online nodes:
+   a) we make use of a flag in the relation data
+   b) we check on the existence of the flag to know if the lock is held or not
+   c) if not there => we set the lock (flag in the peer rel data)
+   d) we release the lock by removing that flag from the rel data
+"""
+import json
+import os
+from typing import List, Optional
+
+import ops
+
+from opensearch_single_kernel.common.client import OpenSearchClient
+from opensearch_single_kernel.common.constants import DeploymentType, Scope, StartMode
+from opensearch_single_kernel.common.exceptions import OpenSearchHttpError
+from opensearch_single_kernel.core.models import DeploymentDescription
+from opensearch_single_kernel.core.state import ClusterState
+from opensearch_single_kernel.managers.base import BaseManager
+from opensearch_single_kernel.utils.helpers import format_unit_name
+from opensearch_single_kernel.utils.topology import ClusterTopology
+from opensearch_single_kernel.workload.base import BaseWorkload
+
+
+class PeerLockManager(BaseManager):
+    """Fallback lock when all units of OpenSearch are offline."""
+
+    def __init__(self, state: ClusterState, workload: BaseWorkload):
+        self.state = state
+        self.workload = workload
+
+    @property
+    def acquired(self) -> bool:
+        """Attempt to acquire lock.
+
+        Returns:
+            Whether lock was acquired
+        """
+        if not self._relation:
+            return False
+
+        self._relation.data[self.state.opensearch_unit.unit]["lock-requested"] = json.dumps(True)
+
+        if self.state.opensearch_unit.is_app_leader:
+            self.logger.debug("[Node lock] Requested peer lock as leader unit")
+            # A separate relation-changed event won't get fired
+            self._on_peer_relation_changed()
+
+        if self._unit_with_lock != self.state.opensearch_unit.unit_name:
+            self.logger.debug(
+                f"[Node lock] Not acquired. Unit with peer databag lock: {self._unit_with_lock}"
+            )
+            return False
+
+        if (
+            self.state.opensearch_unit.is_app_leader
+            and self._relation.data[self.state.opensearch_application.app][
+                "leader-acquired-lock-after-juju-event-id"
+            ]
+            == os.environ["JUJU_CONTEXT_ID"]
+        ):
+            # `unit-with-lock` was set in this Juju event
+            # If the charm code raises an uncaught exception later in the Juju event,
+            # `unit-with-lock` will be reverted to its previous value—which could allow another
+            # unit to get the lock.
+            # Therefore, we cannot use the lock now. We must wait until the next Juju event,
+            # when `unit-with-lock` has been committed (i.e. won't be reverted), to use the
+            # lock.
+            if self.state.opensearch_application.app.planned_units() <= 1:
```

Comment:
```suggestion
           if self.state.planned_units <= 1:
```
self. state.opensearch_application.app does not have planned units.